docs/discovery.md

```diff
@@ -155,3 +102,58 @@ class MyClass
     }
 }
 ```
+
+
+## Integrating your own implementation with the discovery mechanism
+
+The `php-http/discovery` has built-in support for some implementations. To use a different implementation or override 
+the default when several implementations are available in your codebase, you can register a class explicitly with the 
+corresponding discovery service. For example:
+
+``` php
+HttpAdapterDiscovery::register('Acme\MyAdapter`, true);
+```
+
+- `class`: The class that is instantiated. This class MUST NOT require any constructor arguments.
+- `condition`: The condition that is evaluated to boolean to decide whether the class can be instantiated. The following types are allowed:
+    - string: Checked for class existence
+    - callable: Called and evaluated to boolean
+    - boolean: Can be true or false
+    - any other: considered false
+
+The condition can also be omitted. In that case the class is used as the condition (to check for existence).
+
+Classes registered manually are put on top of the list.
+
+
+### Writing your own discovery
+
+Each discovery service is based on the `ClassDiscovery` and has to specify a `cache` field and a `class` field to specify classes for the corresponding service.
```

Comment:
I would add some simple reasoning here: in case of static property inheritance parent property is shared between children. To avoid this, properties MUST be declared in children.